Component API Implementation

0xJoeMama opened the pull request:
After 2 weeks of unhealthy head-banging I have finally finished this!

This API adds:
* A new module to the data library, called 'component'.
* A new class on the qsl_base module, called Maybe. It represents a monad that either holds a value or doesn't.

# A bit of an explanation: 
This module allows you attach data, functions, ticking functions to game objects.
Right now the following objects are implemented:
* Block Entities
* Entities
* Chunks
* ~~ItemStacks~~(Removed from the Initial Release)
* LevelProperties

The data can be serialized using NBT and can also be network synced(although that feature is a bit unstable at the moment).

## Component Injection
During the mod loading stage, mods are able to inject into difference game objects, provided those objects implement the ComponentProvider interface. Those injections are cached using the target class/classes and are then able to be quickly looked up.  

*But what about the case where cached injections don't suffice?* one may ask.
Well that's also covered using **dynamic injection**.  
Along with the class that you are targeting, you can also provide a predicate that will match the type of the class.
That predicate is then used when a provider is initialized to make sure it matches and/or can be injected into the target.
Proceed with caution: **The predicate is checked, at the end of the constructor of the highest class in the hierarchy. This means that only the properties initialized up to that point will be available for querying!.**

While that may not seem like a problem for things like ItemStacks, which generally aren't extended, Entities for example have a wide variety of classes with different fields. However, when injecting dynamically, one will only have access to the Entity class fields as well as the type information of the object.

## Custom implementations
I have made the API really flexible, so that anyone can just create custom implementations of most provided objects.
Component containers, component providers, components, sync headers, all of those can be extended and tweaked as per need.

## Network sync and data saving
Data is currently saved using NBT and a save call that is provided from the provider to its container.
Data is currently synced using a system I have implemented. Needless to say, this system needs some tweaking as well as a lot of testing to make sure that: 
1. No data is lost through the network.
2. ~~There are no runtime crashes caused by the system.~~
3. ~~Components are properly synced even in edge cases.~~
4. ~~Components are synced when the object is first loaded.~~

I think that after all that is done, this will be able to be released.

# Finally
Feel free to ask any questions about the implementation and provide feedback wherever possible.
This project was a massive undertaking for one person and that is why I decided to PR it now that it's reaching its final stages.
The codebase is big and documentation is needed, I understand that. However I still think I needed to share it with everyone.
Thank you for your time!

0xJoeMama committed:
Implemented the core Component API

0xJoeMama committed:
Modified creeper testing a bit

0xJoeMama committed:
Implemented caching for injection targets.

0xJoeMama committed:
Reformat and better tests.

0xJoeMama committed:
Implemented ComponentProvider for Block Entities.

0xJoeMama committed:
Added Chunks are Component Providers(serialization does not work yet).

0xJoeMama committed:
Fully implemented chunk component providers and some itemstack
implementation.

0xJoeMama committed:
Added marking dirty to other providers that require it

0xJoeMama committed:
Fixed a minor issue with chunks not saving qsl data when generated.

0xJoeMama committed:
Made LevelProperties a ComponentProvider allowing you to attach
components to saves.

0xJoeMama committed:
Reformatted code.

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

0xJoeMama committed:
Added some todos.

0xJoeMama committed:
Added container(WIP)

0xJoeMama committed:
Implemented containers fully and switched to using them

0xJoeMama committed:
Added SimpleComponent for custom ComponentProvider implementations.

0xJoeMama committed:
Fixed testmod issue.

0xJoeMama committed:
Added SimpleComponentContainer

0xJoeMama committed:
Turned ComponentProvider into an injected interface.

0xJoeMama committed:
Reformat.

0xJoeMama committed:
Run applyLicenses

0xJoeMama committed:
Added Component.Factory and restructured code.

0xJoeMama committed:
Switched to using vanilla registries to handle our registration process.

0xJoeMama committed:
Woops that went unnoticed.

0xJoeMama committed:
ItemStacks are now ComponentProviders.

0xJoeMama committed:
Removed unneeded imports.

0xJoeMama committed:
Merged implementation of itemstacks.

0xJoeMama committed:
Added component events.

0xJoeMama committed:
Hooked the event for injections properly.

0xJoeMama committed:
Improved API structure.

0xJoeMama committed:
More structure improvements.

0xJoeMama committed:
SimpleComponentContainer builder.

0xJoeMama committed:
Implemented the FunctionComponent library.

* Function components can be called like normal functions provided you
  have access to the container.
* Ticking components are called every tick if their container supports
  it.
* Added builders to the two primary implementations of
  ComponentContainer.
* Added static ComponentTypes. They are basically component singletons.
* Moved some important API to Components.java

0xJoeMama committed:
Finalized the FunctionComponent system.

0xJoeMama committed:
Code reformat

0xJoeMama committed:
Almost the whole implementation of network sync.

0xJoeMama committed:
Finished network sync(still unstable).

0xJoeMama committed:
Apply licenses before PR

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/wrapper/Wrapper.java:
```diff
@@ -0,0 +1,33 @@
+
+package org.quiltmc.qsl.component.api.wrapper;
+
+public class Wrapper<T> {
```
What's the point of this API class? It only seems to be used in your testmod, and there it can easily be replaced by doing this:
```java
final ActionResult[] res = new ActionResult[] { ActionResult.PASS };

OptionalProducingMethod.invoke().ifPresent(thing -> {
    if (thing == ONE)
        res[0] = ActionResult.CONSUME;
})

return res[0];
```

0xJoeMama replied:
Agreed. I had actually forgotten that class even existed to be honest.
The array thing is a bit interesting seeing how I never thought of it. Very nice, very nice!

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentInjectionCache.java:
```diff
@@ -0,0 +1,60 @@
+
+public class ComponentInjectionCache {
+
+	private static ComponentInjectionCache INSTANCE;
```
Unnecessary singleton, just make the fields and methods static.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java:
```diff
@@ -0,0 +1,110 @@
+	private static final Logger LOGGER = LoggerFactory.getLogger("QSL/Components");
+	public static final String MODID = "quilt_component";
+    private static final Map<ComponentInjectionPredicate, Set<Identifier>> INJECTION_REGISTRY = new HashMap<>();
+	private static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
```
This registry key should be exposed.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	@NotNull
+	public static Logger getLogger() {
```
...why not just make the logger `public`?

0xJoeMama replied:
Just a thing I have been doing for all my mods. It looks better in code to me. If you insist though, I have no issue just making the LOGGER field public.

LambdAurora replied:
I would just make the logger field public.

The static method just looks like additional noise to me.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/container/LazifiedComponentContainer.java:
```diff
@@ -0,0 +1,239 @@
+import java.util.*;
+import java.util.function.Supplier;
+
+public class LazifiedComponentContainer implements ComponentContainer {
```
Possibly rename this to `LazyComponentContainer`? "Lazified" makes no sense to me.

0xJoeMama replied:
This was also a bad use of ArrayList. I should've used a Set since the only method called on this collection is 'contains'.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/predicates/FilteredInheritedInjectionPredicate.java:
```diff
@@ -0,0 +1,54 @@
+
+	public FilteredInheritedInjectionPredicate(Class<?> clazz, Class<?>[] exceptions) {
+		super(clazz);
+		this.exceptions = new ArrayList<>(Arrays.asList(exceptions));
```
```suggestion
		this.exceptions = Arrays.asList(exceptions);
```
`Arrays.asList` returns an immutable list. I'm not sure why you'd want this predicate to be mutable, especially since all the other predicates seem to be immutable.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/predicates/RedirectedInjectionPredicate.java:
```diff
@@ -0,0 +1,35 @@
+
+	public RedirectedInjectionPredicate(Class<?> clazz, Set<Class<?>> redirections) {
+		super(clazz);
+		this.redirections = redirections;
```
```suggestion
		this.redirections = Set.copyOf(redirections);
```
Immutability is fun (no it isn't)!

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/util/Lazy.java:
```diff
@@ -0,0 +1,109 @@
+
+	@NotNull
+	public <U> Lazy<U> compose(Function<T, U> transformer) {
+		return Lazy.of(() -> transformer.apply(this.sup.get()));
```
Shouldn't this be this?
```suggestion
		return Lazy.of(() -> transformer.apply(this.get()));
```
Unless bypassing the lazy behavior was intentional.

0xJoeMama replied:
This slipped through the cracks. I don't think that method is even used to be honest. 

*If I am to be truly honest this class was literally copied from this gist of mine: https://gist.github.com/0xJoeMama/314607455836e6bd66e74a3715f23154*

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/util/Lazy.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Override
+	public int hashCode() {
```
...why do these methods exist? I highly doubt someone's gonna use `Lazy`s as map keys.

0xJoeMama replied:
Same as above this was just copied over from my gist without much thought.

Leo40Git commented on library/data/component/src/testmod/java/org/quiltmc/qsl/component/test/TestBlock.java:
```diff
@@ -0,0 +1,96 @@
+		ItemStack handStack = player.getStackInHand(hand);
+
+		if (!handStack.isEmpty()) {
+			final Wrapper<ActionResult> ret = new Wrapper<>(ActionResult.PASS);
+
+			be.expose(ComponentTestMod.CHUNK_INVENTORY).ifPresent(inventoryComponent -> {
```
Replace this with `final ActionResult[] ret = new ActionResult[] { ActionResult.PASS }`, or use `Optional.map`.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java:
```diff
@@ -0,0 +1,110 @@
+
+@ApiStatus.Internal
+public class ComponentsImpl {
+	private static final Logger LOGGER = LoggerFactory.getLogger("QSL/Components");
+	public static final String MODID = "quilt_component";
+    private static final Map<ComponentInjectionPredicate, Set<Identifier>> INJECTION_REGISTRY = new HashMap<>();
+	private static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
+			RegistryKey.ofRegistry(new Identifier("quilt", "components"));
+	public static final Registry<ComponentType<?>> REGISTRY =
+			new SimpleRegistry<>(REGISTRY_KEY, Lifecycle.experimental(), null);
```
Give these declarations some breathing space!
```suggestion
	private static final Logger LOGGER = LoggerFactory.getLogger("QSL/Components");
	
	public static final String MODID = "quilt_component";
	
    private static final Map<ComponentInjectionPredicate, Set<Identifier>> INJECTION_REGISTRY = new HashMap<>();
    
	private static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
			RegistryKey.ofRegistry(new Identifier("quilt", "components"));
	public static final Registry<ComponentType<?>> REGISTRY =
			new SimpleRegistry<>(REGISTRY_KEY, Lifecycle.experimental(), null);
```

0xJoeMama replied:
Okay for this I blame the new intellij version. The new usage fields make it seems like there is space where there actually isn't.

Leo40Git replied:
...I legitimately forgot those existed, I disabled those like day one :P (right-click on them)

Leo40Git reviewed (commented):
...your entire network stack confuses me greatly.

General nitpick: newlines! A lot of these files have a newline after the class declaration for some reason.

Anyway, here's my specific nitpicks.

0xJoeMama commented:
> newlines!  

Is there an issue with leaving a newline after the class declaration?

Leo40Git commented:
Consistency, mostly - none of the current QSL classes have a newline after the class declaration (and heck, even some of the classes in this PR don't have a newline).

0xJoeMama commented:
Okay will do then. Time to mess with IJ settings!

lenrik1589 commented:
I think there is a way to include code style settings in the repo? well, at least for IDEA users

Leo40Git commented:
We do have an [`.editorconfig`](https://github.com/QuiltMC/quilt-standard-libraries/blob/1.19/.editorconfig) file, though it doesn't contain any IJ-specific settings.

lenrik1589 commented:
I see

quat1024 reviewed (commented):
I'm concerned about firing the event for attaching ItemStack components from the ItemStack constructor.

Minecraft Forge does something similar with its capabilities, and it leads to `new ItemStack` and things that call it such as `copy` and `split` being giant performance hazards, as suddenly this fires off an event.

0xJoeMama commented:
> I'm concerned about firing the event for attaching ItemStack components from the ItemStack constructor.
> 
> Minecraft Forge does something similar with its capabilities, and it leads to `new ItemStack` and things that call it such as `copy` and `split` being giant performance hazards, as suddenly this fires off an event.

How would you support dynamic injections in this case though? Technically speaking, you could do a cached injection and only expose the component if the required criteria is met.

0xJoeMama commented:
We could redirect the calls in copy and split though. Hm interesting...

LemmaEOF commented:
Quilt component API 👀 👀 👀 

Right now, there's not really anything to get components off of non-BE blocks/blockstates. This is, of course, a bit tough because blockstates aren't quite instanced in the same way BEs or item stacks are, but I feel like it is important to provide, given that stuff like composters or redstone in vanilla set a convention that you should be able to make something simple without having to use a full block entity. Using a `BlockComponentProvider` interface with access to a full world/pos/state would work, but that makes injection either impossible or *very* unperformant. I might look into making a lookup-like system for this at some point, unless someone else has a better idea.

0xJoeMama commented:
@LemmaEOF That is an interesting idea. However, I'd love to hear how you are planning storing the components. Or are they gonna be static components? (In case you haven't seen the code, static components are basically component singletons)

0xJoeMama committed:
Applied changes requested by Leo40Git.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java:
```diff
@@ -38,60 +38,65 @@
-    private static final Map<ComponentInjectionPredicate, Set<Identifier>> INJECTION_REGISTRY = new HashMap<>();
-	private static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
-			RegistryKey.ofRegistry(new Identifier("quilt", "components"));
+	public static final RegistryKey<Registry<ComponentType<?>>> REGISTRY_KEY =
```
When I said "exposed", I meant in the API (in `Components`, preferably)! Sorry for the confusion.

0xJoeMama replied:
Oh, very well then!

0xJoeMama replied:
![image](https://user-images.githubusercontent.com/46004440/176508281-c817f4a3-c464-40df-a1fc-9592b4746863.png)
However I wish to let you know that, *this^^* is a thing

Leo40Git replied:
Vanilla has separate key fields for all its registries, so whatever

0xJoeMama committed:
Exposed as in part of the public API.

LambdAurora commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/ComponentProvider.java:
```diff
@@ -0,0 +1,54 @@
+})
+public interface ComponentProvider {
+	@NotNull
+	ComponentContainer getContainer();
```
Given that this is an injected interface, we need to make sure that those names don't have too many risks of conflict, so first of all
```suggestion
	ComponentContainer getComponentContainer();
```

call and expose might be fine since they take QSL-specific arguments.

And this method might require a default implementation, as BasiqueEvangelist pointed out multiple times when we dealt with those interfaces, if there's no default implementation it will disrupt development environments and javac when trying to compile a basic Entity, since the issue is those implementations are hidden away via mixin.

LambdAurora commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/ComponentProvider.java:
```diff
@@ -0,0 +1,54 @@
+		return Components.expose(id, this);
+	}
+
+	default void sync() {
```
```suggestion
	default void syncComponent() {
```

BlockEntity API already injects a sync method for BEs.

LambdAurora reviewed (changes requested):
Another global remark: do not use `@NotNull`, we consider that by default everything is not null, and if something might be null we use `@Nullable`.

0xJoeMama committed:
Fixed runtime crash with FilteredInheritedInjectionPredicate

0xJoeMama commented:
I actually mark stuff with NotNull for the kotlin devs out there. 
The kotlin compiler is able to parse NotNull and allow the kotlin side to just use the normal call operator(.) and not the safe-call one(?.).

0xJoeMama committed:
Applied LambdaAurora's requested changes.

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

0xJoeMama committed:
Added the block_entity module as a dependency of the testmod and removed
old code handling it.

LambdAurora commented:
> I actually mark stuff with NotNull for the kotlin devs out there. The kotlin compiler is able to parse NotNull and allow the kotlin side to just use the normal call operator(.) and not the safe-call one(?.).

Might need to talk of this with the Kotlin team then.

Pyrofab commented:
I have not looked at this in too much detail yet, but I would highly discourage attaching a ComponentProvider to item stacks. The ItemStack module was removed in Cardinal Components API because it caused a lot of issues with other mods dealing with ItemStack serialization, had poor performance, required a lot of special casing, and was generally unneeded after the API Lookup API became a thing.

0xJoeMama commented:
> I would highly discourage attaching a ComponentProvider to item stacks

I can understand why that would be an issue. Honestly I run into multiple problems while creating the itemstack implementation of the ComponentProvider interface. ItemStacks are a weird bunch and they are instantiated quite often. Might have to do some intense thinking about this.

0xJoeMama commented:
However I must also say that I'd like to personally experiment with ItemStacks, *because* they are weird.
Who knows lazy instantiation might be a life saver. Or some other performance improvement down the line.
We can call the itemstack implementation experimental even when network sync is fully tested.
I want ItemStacks to work and to be honest even if they don't I'd like to see them fail myself when testing with them.

0xJoeMama commented:
Just not in production.

LambdAurora commented:
Gonna put this here, if something starts blocking out this PR we always can delay said implementation to a later PR so we can do more thinking on it ;)

Sometimes incremental implementations are easier to manage.

0xJoeMama commented:
'said implementation' aka ItemStack, am I correct in that?

LambdAurora commented:
> 'said implementation' aka ItemStack, am I correct in that?

yup, you don't have to get everything at once (and sometimes incremental implementation is easier to review too ahah :p)

0xJoeMama commented:
If we are talking safety, then Entity and BlockEntity could literally be released right now.
Chunks needs a bit work, to do with initial syncing.
LevelProperties aka 'Level' needs a way to encode itself safely.
ItemStacks are technically not synced through network on our part. That is because we handle ItemStack component nbt under the normal 'nbt' fields in their class. Since that field is synced through PacketByteBuf#writeItemStack we needn't worry about network sync. Issues like frequent instantiation are what are holding ItemStacks back.

0xJoeMama commented:
> yup, you don't have to get everything at once (and sometimes incremental implementation is easier to review too ahah :p)

'We can call the itemstack implementation experimental even when network sync is fully tested.'
Literally my words. I couldn't agree more.

0xJoeMama commented:
> easier to review too

By the way, I did think about how much hell reviewing this PR would be. 
Sorry everyone for probably taking a good chunk of time from your day.

LambdAurora commented:
> > easier to review too
> 
> By the way, I did think about how much hell reviewing this PR would be. Sorry everyone for probably taking a good chunk of time from your day.

It's fine, the API looks cool, I haven't yet jumped too much into reviewing but from afar it's cool stuff!

0xJoeMama commented:
> from afar it's cool stuff!

It is. And I think I've made it quite powerful. You see, I'm the simple by default, powerful when needed kind of person when it comes to APIs(and other things but that's not relevant here).

0xJoeMama committed:
Made the LOGGER fields public and removed the getLogger method.

0xJoeMama committed:
Switched to IdentityHashMap and Component type keys.

0xJoeMama committed:
Added GenericComponent

0xJoeMama committed:
Removed FunctionComponents after discussion.

0xJoeMama committed:
Minor changes here and there.

0xJoeMama committed:
Added OnAccessComponentContainer and restructured a bit.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/sync/header/SyncPacketHeader.java:
```diff
@@ -17,33 +17,38 @@
 		buf.writeInt(SyncHeaderRegistry.HEADERS.getRawId(this));
-		//noinspection unchecked the person calling is responsible to make sure we get a valid provider instance!
+		// the person calling is responsible to make sure we get a valid provider instance!
+		//noinspection unchecked
```
`//noinspection` comments only work in IntelliJ! Annotate this method with `@SuppressWarnings("unchecked")` instead.

Leo40Git commented:
I feel as if `Lazy` should be moved to `qsl_base` as a public API class... but that's out of scope for this PR.

0xJoeMama commented:
> I feel as if `Lazy` should be moved to `qsl_base` as a public API class... but that's out of scope for this PR.

Well if we are talking moving things out, then I'd suggest NetworkCodec be moved to the networking API. It's actually really useful.

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/sync/codec/NetworkCodec.java:
```diff
@@ -0,0 +1,122 @@
+					.map(chunkPos -> MinecraftClient.getInstance().world.getChunk(chunkPos.x, chunkPos.z))
+					.orElseThrow()
+	);
+	public static final NetworkCodec<ItemStack> ITEMSTACK = new NetworkCodec<>(
```
```suggestion
	public static final NetworkCodec<ItemStack> ITEM_STACK = new NetworkCodec<>(
```
Correct conversion between class name (`PascalCase`) and `static final` field name (`SCREAMING_SNAKE_CASE`).

Leo40Git commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/sync/codec/NetworkCodec.java:
```diff
@@ -0,0 +1,122 @@
+			buf -> ClientSyncHandler.getInstance().getType(buf.readInt())
+	);
+
+	public static <O, L extends List<O>> NetworkCodec<L> list(NetworkCodec<O> entryCodec, Function<Integer, L> function) {
```
```suggestion
	public static <O, L extends List<O>> NetworkCodec<L> list(NetworkCodec<O> entryCodec, Function<Integer, L> listFactory) {
```
Or alternatively introduce a custom interface for this:
```java
@FunctionalInterface
public interface ListFactory<E> {
    List<E> createList(int size);
}
```

0xJoeMama committed:
Applied changed requested from Leo40Git

0xJoeMama committed:
Added the Maybe monad class to the base QSL module.

0xJoeMama committed:
Major rewrite of injections and removed notnulls.

0xJoeMama committed:
Removed ItemStacks from ComponentProviders.
* Various other fixes.

0xJoeMama committed:
Cleaned up networking code.

0xJoeMama committed:
Reverted small mistake.

LemmaEOF commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -0,0 +1,517 @@
+ * @param <T> The type of the wrapped object
+ * @author 0xJoeMama
+ */
+public sealed abstract class Maybe<T> permits Maybe.Just, Maybe.Nothing {
```
oh right, sealed classes got finalized! very nice use of them ^.^

0xJoeMama replied:
Wait, are they not finalized? Is there any issue with this?
I could resolve them by making the Maybe constructor private if there is some issue.

LemmaEOF replied:
they are finalized now! I just forgot that they were because I hadn't seen them used yet lol, this is more than fine

0xJoeMama replied:
_Not gonna lie, I thought you were about to yeet out my monad_

EnnuiL commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -0,0 +1,517 @@
+		 */
+		private static final Nothing<?> INSTANCE = new Nothing<>();
+
+		private Nothing() {
```
Nitpick: Drop the line break, turn this into `{ }`

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/ComponentType.java:
```diff
@@ -0,0 +1,68 @@
+	public static class Static {
+		private final Map<ComponentType<?>, Component> staticInstances = new IdentityHashMap<>();
+
+		private Static() {
```
Nitpick: Turn this into a `{ }`, economize lines

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/ComponentType.java:
```diff
@@ -0,0 +1,68 @@
+		}
+
+		@SuppressWarnings("unchecked")
+		<C extends Component> C getOrCreate(ComponentType<C> type, Runnable saveOperation, Runnable syncOperation) {
```
The visibility marker (...? I don't remember the name) should be explicit on here

Leo40Git replied:
Visibility modifier, IIRC.

And not having one _is_ valid - it means the method is _package-private_, and can only be accessed by classes in the same package.

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/components/FloatComponent.java:
```diff
@@ -0,0 +1,26 @@
+package org.quiltmc.qsl.component.api.components;
+
+import org.quiltmc.qsl.component.api.Component;
+
```
Too many lines
```suggestion
```

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/components/NbtComponent.java:
```diff
@@ -0,0 +1,69 @@
+			case NbtElement.LONG_TYPE -> ((NbtComponent<NbtLong>) nbtComponent).read((NbtLong) nbtTarget);
+			case NbtElement.FLOAT_TYPE -> ((NbtComponent<NbtFloat>) nbtComponent).read((NbtFloat) nbtTarget);
+			case NbtElement.DOUBLE_TYPE -> ((NbtComponent<NbtDouble>) nbtComponent).read((NbtDouble) nbtTarget);
+			case NbtElement.BYTE_ARRAY_TYPE ->
```
The line breaks here seems way too arbitrary on this wall of text (oh, and on a kinda unrelated note: I love switch expressions);
Yes yes, there's a char limit per line and such; Ignore 'em here, remove the line breaks on BYTE_ARRAY_TYPE and LONG_ARRAY_TYPE; These limits do make sense as a style guide and you should follow them when it makes sense, but strictly enforcing them just because it passed few characters from it doesn't make sense (heck, iiirc, Linus Torvalds complained about the strictness of that)

0xJoeMama replied:
IntelliJ autoformatting got me on those. I actually just had them all be one liners.

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/CommonInitializer.java:
```diff
@@ -0,0 +1,47 @@
+
+@ApiStatus.Internal
+public final class CommonInitializer implements ModInitializer {
+	public static final String MODID = "quilt_component";
```
Rename MODID to MOD_ID
```suggestion
	public static final String MOD_ID = "quilt_component";
```

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/ComponentsImpl.java:
```diff
@@ -0,0 +1,74 @@
+	public static final CachedInjectionManager CACHED_MANAGER = new CachedInjectionManager();
+	public static final DynamicInjectionManager DYNAMIC_MANAGER = new DynamicInjectionManager();
+
+	public static final Logger LOGGER = LoggerFactory.getLogger("QSL/Components");
```
While there is no standard on the logger names, we can at least begin to standardize to one
```suggestion
	public static final Logger LOGGER = LoggerFactory.getLogger("Quilt Component");
```
This reminds me, i should get everything else to this format

EnnuiL commented on library/data/component/build.gradle:
```diff
@@ -0,0 +1,70 @@
+}
+
+qslModule {
+	name = "Quilt Component Attachment API"
```
Hmm, the divergence between the mod name (Quilt Component Attachment API) and the mod ID (quilt_component) is a bit weird here, and thinking about it, "Component Attachment" doesn't really make sense compared to "Registry Entry Attachments", since for the latter, registry attachments basically attach values to their registry entries, while components aren't the thing being attached on

EnnuiL commented on library/data/component/build.gradle:
```diff
@@ -0,0 +1,70 @@
+	name = "Quilt Component Attachment API"
+	moduleName = "component"
+	id = "quilt_component"
+	description = "API allowing attachment and querying of an ECS-like interface for entities, block-entities, chunks, etc."
```
Hmm, this description needs a bit of improvement, although I'm leaving this for my second look later

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/provider/ComponentProvider.java:
```diff
@@ -0,0 +1,49 @@
+import org.quiltmc.qsl.component.api.Components;
+import org.quiltmc.qsl.component.impl.container.AbstractComponentContainer;
+
+@InjectedInterface({
```
The double tabs should be just one here

Future Me here: oh god, there's a heckton of double tabs everywhere; They all (or at least the majority of them) should be reduced to a single one

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/container/AbstractComponentContainer.java:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	public record ContainerOperations(Runnable saveOperation,
+									  Function<ComponentType<?>, Runnable> syncOperationFactory) {
```
eugh, this thing should be in a single line

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/container/EmptyComponentContainer.java:
```diff
@@ -0,0 +1,60 @@
+	public static final ComponentContainer.Factory<EmptyComponentContainer> FACTORY =
+			(provider, injections, saveOperation, ticking, syncContext) -> INSTANCE;
+
+	private EmptyComponentContainer() {
```
Turn this (and the below empty methods) into a `{ }`

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/container/LazyComponentContainer.java:
```diff
@@ -0,0 +1,92 @@
+
+	@Override
+	protected <COMP extends Component> void addComponent(ComponentType<COMP> type, Component component) {
+
```
`{ }`

EnnuiL commented on library/data/component/src/main/java/org/quiltmc/qsl/component/impl/util/StringConstants.java:
```diff
@@ -0,0 +1,21 @@
+package org.quiltmc.qsl.component.impl.util;
+
+public final class StringConstants {
+	public static final String COMPONENT_ROOT = "QslComponentRoot";
```
```suggestion
	public static final String COMPONENT_ROOT = "QuiltComponentRoot";
```

EnnuiL commented on library/data/component/src/testmod/resources/quilt_component_test.mixins.json:
```diff
@@ -0,0 +1,15 @@
+  "package": "org.quiltmc.qsl.component.test.mixin",
+  "compatibilityLevel": "JAVA_17",
+  "minVersion": "0.8",
+  "mixins": [
```
This field can be removed

0xJoeMama replied:
I suppose you mean minVersion, right?

EnnuiL replied:
No, I mean the empty mixins field; minVersion is fine

EnnuiL commented on library/data/component/src/testmod/java/org/quiltmc/qsl/component/test/client/TestBeRenderer.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.util.function.Predicate;
+
+public class TestBeRenderer implements BlockEntityRenderer<TestBlockEntity> {
```
This should be named `TestBlockEntityRenderer`

EnnuiL reviewed (changes requested):
I'll do my review into two passes, since yeah, there's a lot to see and to be done;

On this first pass, yeah, it's pretty much style and standarization issues; I recommend using the project's Checkstyle config in order to clean up the rest of them, 'cause yeah, this isn't an exaustive one
As for the logic itself? On a glance, all seems great! Although I'm curious about whenever there are holes to poke on or not, and yeah, I'll have to actually mess around with code and such for that

0xJoeMama commented:
Oh, by the way. Dynamic injections don't use the event system anymore.

0xJoeMama committed:
Applied changes requested by EnnuiL

Leo40Git commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -0,0 +1,516 @@
+	 * Similar function to {@link Optional#empty()}.
+	 *
+	 * @param <T> The type the {@link Nothing} instance will take the form of.
+	 * @return the singleton {@link Nothing} instance after casting it to the proper type.
```
```suggestion
	 * @return A {@link Nothing} instance.
```
Exposing this implementation details is just gonna make people do `maybe == Maybe.nothing()`.

Leo40Git commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -0,0 +1,516 @@
+	 * @see Optional#orElseGet
+	 */
+	public abstract T unwrapOrGet(Supplier<? extends T> supplier);
+
```
Maybe add an `unwrapOrThrow(Supplier<? extends Throwable> exceptionSupplier)` method?

Leo40Git commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -0,0 +1,516 @@
+	public abstract T unwrapOr(T defaultValue);
+
+	/**
+	 * Works like {@link Maybe#unwrapOr} except that it lazily evaluated the return value using the provided {@link Supplier}.
```
```suggestion
	 * Works like {@link Maybe#unwrapOr} except that it lazily evaluated the return value using the provided {@link Supplier}.
```
```suggestion
	 * Works like {@link Maybe#unwrapOr} except that it lazily evaluates the return value using the provided {@link Supplier}.
```

0xJoeMama committed:
Applied suggestions Leo40Git

Leo40Git commented on library/core/qsl_base/src/main/java/org/quiltmc/qsl/base/api/util/Maybe.java:
```diff
@@ -267,6 +267,15 @@ public static <T> Maybe<T> fromOptional(Optional<T> opt) {
+	 * @return The contained value, if it exists.
+	 * @throws Throwable A customizable throwable.
+	 */
+	public abstract T unwrapOrThrow(Supplier<? extends Throwable> throwableSupplier) throws Throwable;
```
```suggestion
	public abstract <E extends Throwable> T unwrapOrThrow(Supplier<? extends E> throwableSupplier) throws E;
```
This prevents checked exceptions from being thrown without javac realizing (or at least, makes it harder to do this accidentally).

0xJoeMama committed:
Applied Leo40Git's suggestions + other minor changes.

0xJoeMama committed:
Major Rewrite + Docs(doesn't function right now)

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

0xJoeMama committed:
Cleared up some code.

0xJoeMama committed:
Rewrite of syncing:
* Implemented initial sync.
* Implemented sync requests.

This is the last feature commit. From now on only improvements, until
the initial release.

0xJoeMama commented:
Consider the initial feature-set done! Feel free to start breaking this!

0xJoeMama committed:
Reformat using Checkstyle

0xJoeMama committed:
Run applyLicenses

0xJoeMama committed:
Removed the Component interface.

0xJoeMama committed:
Fixed docs.

0xJoeMama committed:
Revamped the way to perform injections.

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

oliviathevampire commented on library/data/component/src/main/resources/quilt_component.mixins.json:
```diff
@@ -0,0 +1,25 @@
+  "package": "org.quiltmc.qsl.component.mixin",
+  "compatibilityLevel": "JAVA_17",
+  "mixins": [
+    "RegistryAccessor",
```
This errors? Does this mixin even exist?

0xJoeMama replied:
Woops my bad!

0xJoeMama committed:
Removed invalid mixin

0xJoeMama committed:
Fixed various issues and resolved some todoes

oliviathevampire commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/provider/ComponentProvider.java:
```diff
@@ -0,0 +1,92 @@
+ * needs to implement this interface.<br/>
+ *
+ * <p>
+ * Currectly it consists of the {@link ComponentProvider#getComponentContainer()} method, which is the abstract one,
```
``Currently,`` instead of ``Currectly``

oliviathevampire commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/provider/ComponentProvider.java:
```diff
@@ -0,0 +1,92 @@
+ * @see ComponentContainer
+ * @see Components
+ */
+@InjectedInterface({ // We inject this inteface, so that modders don't need to use the methods in Components directly with our default implementations.
```
``interface`` instead of ``inteface``

oliviathevampire commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/provider/ComponentProvider.java:
```diff
@@ -0,0 +1,92 @@
+public interface ComponentProvider {
+	/**
+	 * Every {@linkplain ComponentProvider provider} must provide a {@link ComponentContainer},
+	 * so that it can store the components targetting or that are manually added to it.
```
``targeting`` instead of ``targetting``

0xJoeMama committed:
Component injections are now the responsibility of ComponentContainer.Builders

0xJoeMama committed:
Added the default component container factory that will be used.

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

lukebemish commented on library/data/component/src/main/java/org/quiltmc/qsl/component/api/provider/ComponentProvider.java:
```diff
@@ -0,0 +1,92 @@
+	 * @param <C>  The type of the held component.
+	 * @return A {@link Maybe} instance following the rules defined in {@link Components#expose(ComponentType, Object)}.
+	 */
+	default <C> Maybe<C> expose(ComponentType<C> type) {
```
Is there a particular benefit to exposing a `Maybe`/`Optional` here as opposed to simply returning a `@Nullable` value? My initial reaction, especially given that this method could be run quite often and constructing lots of objects could add some cost (though I'm not actually sure how bad this would be, since I haven't benchmarked it or anything, so this could be premature optimization) is that it's better to return a nullable value. However, I may be missing something regarding this.

0xJoeMama replied:
In the default implementations of ComponentContainer, the optional object is cached and so no object is created except the one at the initialization of the container.

lukebemish replied:
Ah. I meant to leave this comment on `ComponentContainer#expose`, which your javadocs note is called *really* often

0xJoeMama replied:
It's okay, this method also propagates to that method. In other words the same applies here as well.

lukebemish replied:
Yes. Personally, I'd propose switching both to just be nullable and cut out Maybe entirely.

0xJoeMama replied:
Well, thing is the reason it's Optional in the first place is for API's sake. Optional code is much cleaner and understandable to look at, than nullable code in my opinion. However, there is the solution of adding utility methods in the ComponentProvider interface, that will make the code cleaner. I'll give those a try, and see what it'll look like.

Thanks for the suggestion!

0xJoeMama committed:
Merge branch 'QuiltMC:1.19' into 1.19

0xJoeMama committed:
Nullificating Part 01

0xJoeMama committed:
Nullification Part 02
